### PR TITLE
feat(*) Remove jitter with plural events and expected events

### PR DESCRIPTION
**Pull request by brooksmtownsend**
There's one or two more TODOs that I've outlined below, but I'm marking as ready for review to get some input around the rest of the methodology. 

This PR:
- [x] Adds support for ActorsStarted / ActorsStopped and their failure events
- [x] Implicitly requires `wadm` to work with v0.63+ versions of the host because of the above
- [x] Removes the notion of a backoff, replacing it with logic to make Scalers compute "expected events" for each command that they send. For example, if a scaler wants to start 5 of an actor, it won't take any action until it either receives an event where 5 of that actor started, or if that command failed to complete. Scaler implementers do not need to care about this, it's all on the `wadm` side.
- [x] TODO: Make the expected events list clear after a certain time frame to account for missed events or strange edge cases.

## Feature or Problem
Currently, Scalers (particularly the ActorSpreadScaler) are prone to jitter and difficulty actually converging on proper state when you run many (10+) copies of an actor. They are handling each event as it comes in, and we added a very coarse backoff mechanism to account for this but it still creates difficulties.

## Related Issues
Will file issue soon to make state only update on plural events.

## Release Information
next alpha release

## Consumer Impact
Consumers will notice that wadm feels more responsive at reconciling

## Testing

Built on platform(s)
- [ ] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [x] aarch64-darwin
- [ ] x86_64-windows

Tested on platform(s)
- [ ] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [x] aarch64-darwin
- [ ] x86_64-windows

### Unit Test(s)
This feature is large enough that it probably warrants some tests to verify behavior around expected events.

- [ ] TODO: Tests

### Manual Verification
I've applied applications and messed with them in many ways to confirm that this works great with 1 running wadm.

### CODE REVIEW

File: src/scaler/spreadscaler/mod.rs [hunk 1]
```diff
@@ -62,15 +61,14 @@ impl<S: ReadStore + Send + Sync + Clone> Scaler for ActorSpreadScaler<S> {
         };
         self.config.spread_config = spread_config;
         self.spread_requirements = compute_spread(&self.config.spread_config);
+        //TODO: do we need to remove the backoff?
```
Comment (thomastaylor312) [inline src/scaler/spreadscaler/mod.rs:64]: I assume yes at this point? 😄

File: src/workers/event.rs [hunk 2]
```diff
@@ -701,53 +701,90 @@ where
...
+///
+/// # Returns
+///
+/// * `Result<Vec<Command>>`: A `Result` containing a vector of `Command` structs if successful,
+///   or an error of type `anyhow::Error` if any error occurs while processing the event.
+async fn scaler_handle_event(scaler: &ScalerWithEvents, event: &Event, model_name: &str) -> anyhow::Result<Vec<Command>> {
+    // println!("Handling event {:?}", event);
+
+    // If a scaler is expecting events still, don't have it handle events
+    //TODO: we may need to expire this after a time period.
```
Comment (thomastaylor312) [inline src/workers/event.rs:761]: Yep, this should clear out after a while

**Review by thomastaylor312 — COMMENTED** [review]
Some initial comments. Might want to schedule a synchronous review on this one with you

File: src/commands/mod.rs [hunk 3]
```diff
@@ -7,6 +7,14 @@ use std::{
 
 use serde::{Deserialize, Serialize};
 
+use crate::{
+    events::{
+        ActorClaims, ActorsStartFailed, ActorsStarted, Event, Linkdef, LinkdefSet, ProviderClaims,
+        ProviderStartFailed, ProviderStarted,
+    },
+    workers::insert_managed_annotations,
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply src/commands/mod.rs:15]: Moved to event_helpers in workers, I actually don't mind it there. Would be fine considering moving though if you have another suggestion

File: .devcontainer/install.sh [hunk 4]
```diff
@@ -1,20 +1,5 @@
...
-TARBALL=wadm-$VERSION-linux-$ARCH
-
-curl -fLO https://github.com/wasmCloud/wadm/releases/download/$VERSION/$TARBALL.tar.gz
-tar -xvf $TARBALL.tar.gz
-chmod +x $TARBALL/wadm
-mv $TARBALL/wadm /usr/local/cargo/bin/wadm
-rm -rf $TARBALL $TARBALL.tar.gz
\ No newline at end of file
+curl -s https://packagecloud.io/install/repositories/wasmcloud/core/script.deb.sh | sudo bash
+sudo apt install wash openssl -y
```
Comment (brooksmtownsend) [inline .devcontainer/install.sh:5]: Note: devcontainer only takes like 30 seconds to startup now!

**Comment by brooksmtownsend** [comment]
Pretty positive I handled the rebase properly, but if I removed something from #118 lmk

File: src/events/types.rs [hunk 5]
```diff
@@ -83,10 +83,14 @@ pub trait EventType {
 }
 
 /// A lattice event
-#[derive(Debug)]
+// TODO:  this `Deserialize` impl _compiles_ but doesn't let you deserialize an event because of our custom Serialize
```
Comment (brooksmtownsend) [inline src/events/types.rs:86]: @thomastaylor312 just a note, this is probably why tests fail and all. The notifications that we try to send to remove an event or register events fail to deserialize on the other end because they use the derived `Deserialize` but the manually implemented `Serialize`

File: src/commands/mod.rs [hunk 6]
```diff
@@ -30,6 +37,127 @@ pub enum Command {
...
+                        public_key: actor_id.to_owned(),
+                        host_id: host_id.to_owned(),
+                        count: *count,
+                        // We don't know if there are additional instances, so can't consider remaining
+                        remaining: 0,
+                    }),
+                    None,
+                ))
+            }
+            Command::StartProvider(StartProvider {
```
Comment (connorsmith256) [inline src/commands/mod.rs:101]: No `StopProvider` or `DeleteLinkdef`?

File: src/events/types.rs (hunk 5 above)
Comment (connorsmith256) [reply src/events/types.rs:89]: @thomastaylor312 is this still an issue?

File: src/scaler/manager.rs [hunk 7]
```diff
@@ -32,108 +31,77 @@ use crate::{
...
 
 /// All events sent for manifest notifications
 #[derive(Debug, Serialize, Deserialize)]
 pub enum Notifications {
     CreateScalers(Manifest),
     DeleteScalers(String),
-    BackoffScaler(String),
+    /// Register expected events for a manifest. You can either trigger this with an event (which
+    /// will result in calling `handle_event`) or without in order to calculate expected events with
+    /// a full reconcile. is being done (like on first deploy), rather than just handling a single
```
Comment (connorsmith256) [inline src/scaler/manager.rs:54]: > is being done

Might have deleted something?

File: src/scaler/manager.rs [hunk 8]
```diff
@@ -298,32 +262,35 @@ where
...
+    pub async fn get_specific_scaler<'a>(
+        &'a self,
+        name: &'a str,
+        scaler_id: &'a str,
+    ) -> Option<SingleScaler> {
+        let lock = self.scalers.clone().read_owned().await;
+        OwnedRwLockReadGuard::try_map(lock, |scalers| match scalers.get(name) {
+            Some(scalers) => scalers.iter().find(|s| s.id() == scaler_id),
+            None => None,
+        })
```
Comment (connorsmith256) [inline src/scaler/manager.rs:284]: Supernit: does `and_then` work?
```rust
scalers
    .get(name)
    .and_then(|scalers| scalers.iter().find(|s| s.id() == scaler_id)
```

File: src/scaler/manager.rs [hunk 9]
```diff
@@ -441,16 +406,43 @@ where
...
+                                                Ok(e) => e,
+                                                Err(e) => {
+                                                    error!(error = %e, %name, "Unable to parse given event");
+                                                    continue;
+                                                }
+                                            };
+                                            if let Err(e) = scaler.handle_event(&parsed_event).await {
+                                                error!(error = %e, %name, %scaler_id, "Unable to register expected events for scaler");
+                                            }
+                                        } else if let Err(e) = scaler.reconcile().await {
```
Comment (connorsmith256) [inline src/scaler/manager.rs:424]: Not sure I follow this. Why do we call `reconcile` if there's no triggering event? (And why do we not call reconcile when removing an expected event?)

File: src/scaler/mod.rs [hunk 10]
```diff
@@ -39,18 +60,363 @@ pub trait Scaler {
...
+    async fn remove_event(&self, event: &Event) -> Result<bool> {
+        let mut expected_events = self.expected_events.write().await;
+        let before_count = expected_events.len();
+        expected_events.retain(|(success, fail)| {
+            !evt_matches_expected(success, event)
+                && !fail
+                    .as_ref()
+                    .map(|f| evt_matches_expected(f, event))
+                    .unwrap_or(false)
+        });
```
Comment (connorsmith256) [inline src/scaler/mod.rs:141]: This is hard to follow. Are we removing if the event matches _either_ the success or fail in any of the `expected_events`?

File: src/scaler/mod.rs [hunk 11]
```diff
@@ -39,18 +60,363 @@ pub trait Scaler {
...
+            trace!("Scaler is not backing off, handling event");
+            let commands = self.scaler.handle_event(event).await?;
+
+            // Based on the commands, compute the events that we expect to see for this scaler. The scaler
+            // will then ignore incoming events until all of the expected events have been received.
+            let expected_events = commands
+                .iter()
+                .filter_map(|cmd| cmd.corresponding_event(model_name));
+
+            // Only let other scalers know if we generated commands to take
```
Comment (connorsmith256) [inline src/scaler/mod.rs:192]: > if we generated commands to take

To take what?

File: src/scaler/mod.rs [hunk 12]
```diff
@@ -39,18 +60,363 @@ pub trait Scaler {
...
+            // NOTE(brooksmtownsend): I tried to make the events iterator peekable and look at that instead
+            // but that resulted in a "higher ranked lifetime error". Weird.
```
Comment (connorsmith256) [inline src/scaler/mod.rs:194]: Could link to the rust issue with a TODO to revisit once that's fixed?

File: src/scaler/mod.rs [hunk 13]
```diff
@@ -39,18 +60,363 @@ pub trait Scaler {
...
+        // Clear any existing handle
+        if let Some(handle) = event_cleaner.take() {
+            handle.abort();
+        }
+        let expected_events = self.expected_events.clone();
+        let timeout = self.cleanup_timeout;
+
+        *event_cleaner = Some(tokio::spawn(async move {
+            tokio::time::sleep(timeout).await;
+            // We don't use the method to clear here because of borrowing semantics.
```
Comment (connorsmith256) [inline src/scaler/mod.rs:263]: I don't understand this comment, since we are calling `clear` on the next line

File: src/scaler/mod.rs [hunk 14]
```diff
@@ -39,18 +60,363 @@ pub trait Scaler {
...
+            }
+            Ok(commands) => {
+                trace!("Reconcile generated no commands, no need to register expected events");
+                Ok(commands)
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Sets a timed cleanup task to clear the expected events list after a timeout
```
Comment (connorsmith256) [inline src/scaler/mod.rs:251]: Why is this needed?

File: src/workers/event.rs [hunk 15]
```diff
@@ -853,6 +864,39 @@ where
...
+    Fut: futures::Future<Output = Result<Vec<Command>>>,
+    I: IntoIterator<Item = Fut>,
+{
+    let (commands, errors): (Vec<_>, Vec<_>) = futures::future::join_all(futs)
+        .await
+        .into_iter()
+        .partition(Result::is_ok);
+    // SAFETY: We partitioned by Ok and Err, so the unwraps are fine
+    (
+        commands.into_iter().flat_map(Result::unwrap).collect(),
```
Comment (connorsmith256) [inline src/workers/event.rs:880]: I assume this is right, but having a hard time parsing why it's `flat_map` instead of `map`

**Review by connorsmith256 — COMMENTED** [review]
Mostly questions on my part

File: src/events/types.rs (hunk 5 above)
Comment (thomastaylor312) [reply src/events/types.rs:89]: Nope, all fixed now, forgot to remove todo

File: src/commands/mod.rs (hunk 6 above)
Comment (thomastaylor312) [reply src/commands/mod.rs:101]: I'm not sure what @brooksmtownsend was going for here, but my guess is that both of those don't matter for backing off. If the scaler was the one that stopped a provider, that is fine as it will get picked up by the scaler, see that it doesn't need to reconcile, and continue on. This is due to only one provider being able to run per link name + host combo.

The linkdef is similar, if we want to delete it, then it doesn't matter when we receive it as there can only ever be one linkdef (per link, provider, actor).

File: src/scaler/manager.rs (hunk 8 above)
Comment (thomastaylor312) [reply src/scaler/manager.rs:284]: Why yes it does

File: src/scaler/manager.rs (hunk 9 above)
Comment (thomastaylor312) [reply src/scaler/manager.rs:424]: This is cheating and I'll add a comment to clarify. Basically, if this is a backoff wrapped scaler, it will cause the expected events to be added to the list. If it isn't we just return the commands and do nothing with them

File: src/scaler/mod.rs (hunk 10 above)
Comment (thomastaylor312) [reply src/scaler/mod.rs:141]: Yes. Any command we send out could have either a success case or an optional error case. Matching either one means we found one of the expected events

File: src/scaler/mod.rs (hunk 14 above)
Comment (thomastaylor312) [reply src/scaler/mod.rs:251]: We don't want to stay in a backoff state forever, this handles the case of a network interruption or missed events by eventually expiring them

File: src/workers/event.rs (hunk 15 above)
Comment (thomastaylor312) [reply src/workers/event.rs:880]: The inner type of each result is `Vec<Command>`, so this shortcuts it from `commands.into_iter().map(Result::unwrap).flatten().collect()`. I think it was clippy who told me not to do that

File: src/scaler/mod.rs (hunk 13 above)
Comment (thomastaylor312) [reply src/scaler/mod.rs:263]: I think it was left over from something before the refactor

File: src/scaler/mod.rs (hunk 12 above)
Comment (thomastaylor312) [reply src/scaler/mod.rs:194]: I just removed this for now (cc @brooksmtownsend), I wasn't sure what he was trying to do with peekable. Can add back in later

File: src/scaler/mod.rs (hunk 10 above)
Comment (connorsmith256) [reply src/scaler/mod.rs:141]: It took me a minute to grok what was happening. Mind adding a comment there?

**Review by connorsmith256 — APPROVED** [review]
